pkg/container/qemu_runner.go

```diff
@@ -0,0 +1,879 @@
+			}
+			f.Close()
+		case tar.TypeSymlink:
+			if err := os.Symlink(hdr.Linkname, filepath.Join(guestDir, hdr.Name)); err != nil {
+				return ref, fmt.Errorf("failed to create symlink %s: %w", fullname, err)
+			}
+		case tar.TypeLink:
+			if err := os.Link(filepath.Join(guestDir, hdr.Linkname), filepath.Join(guestDir, hdr.Name)); err != nil {
+				return ref, fmt.Errorf("failed to create hardlink %s: %w", fullname, err)
+			}
+		default:
+			// TODO: Is this correct? We are loading these into the directory, so character devices and such
+			// do not really matter to us, but maybe they should?
+			continue
+		}
+	}
+	return guestDir, nil
+}
+
+func (b qemuOCILoader) RemoveImage(ctx context.Context, ref string) error {
+	clog.FromContext(ctx).Infof("removing image path %s", ref)
+	return os.RemoveAll(ref)
+}
+
+func createMicroVM(ctx context.Context, cfg *Config) error {
+	rootfs := cfg.ImgRef
+	baseargs := []string{}
+
+	// always be sure to create the VM rootfs first!
+	kernelPath, rootfsInitrdPath, err := createRootfs(ctx, cfg, rootfs)
+	if err != nil {
+		clog.FromContext(ctx).Errorf("could not prepare rootfs: %v", err)
+		return err
+	}
+
+	// load microvm profile and bios, shave some milliseconds from boot
+	// using this will make a complete boot->initrd (with working network) In ~700ms
+	// instead of ~900ms.
+	if _, err := os.Stat("/usr/share/qemu/bios-microvm.bin"); err == nil {
+		// only enable pcie for network, enable RTC for kernel, disable i8254PIT, i8259PIC and serial port
+		baseargs = append(baseargs, "-machine", "microvm,rtc=on,pcie=on,pit=off,pic=off,isa-serial=off")
+		baseargs = append(baseargs, "-bios", "/usr/share/qemu/bios-microvm.bin")
+	}
+
+	if cfg.Memory != "" {
+		baseargs = append(baseargs, "-m", cfg.Memory)
+	} else {
+		baseargs = append(baseargs, "-m", getAvailableMemoryKB())
+	}
+
+	if cfg.CPU != "" {
+		baseargs = append(baseargs, "-cpu", cfg.CPU)
+	} else {
+		baseargs = append(baseargs, "-cpu", "host")
+	}
+
+	// use kvm on linux, and Hypervisor.framework on macOS
+	if runtime.GOOS == "linux" {
+		baseargs = append(baseargs, "-enable-kvm")
+	} else if runtime.GOOS == "darwin" {
+		baseargs = append(baseargs, "-accel", "hvf")
+	}
+
+	baseargs = append(baseargs, "-smp", fmt.Sprintf("%d", runtime.NumCPU()))
+	baseargs = append(baseargs, "-daemonize")
+	// ensure we disable unneeded devices, this is less needed if we use microvm machines
+	// but still useful otherwise
+	baseargs = append(baseargs, "-display", "none")
+	baseargs = append(baseargs, "-no-reboot")
+	baseargs = append(baseargs, "-no-user-config")
+	baseargs = append(baseargs, "-nodefaults")
+	baseargs = append(baseargs, "-parallel", "none")
+	baseargs = append(baseargs, "-serial", "none")
+	baseargs = append(baseargs, "-vga", "none")
+	// use -netdev + -device instead of -nic, as this is better supported by microvm machine type
+	baseargs = append(baseargs, "-netdev", "user,id=id1,hostfwd=tcp::"+cfg.SSHPort+"-:22")
```

Comment:
https://pkg.go.dev/github.com/mdlayher/vsock would maybe allow us to use a vsock for host<->guest . I guess you still need user networking for host -> internet.

Reply:
That's something to explore thanks
This would remove that ~5s delay of the `sshd` starting up and would help a bit for smaller packages in proportion

Reply:
In the meantime I've improved the sshd startup time, improving the entropy (passing the host's one) and properly checking for the server to come up (instead of using ssh client that is way slower)

Reply:
+1 on the hosts entropy.